### PR TITLE
housekeeping: prevent log warning for missing getComponent in production

### PR DESCRIPTION
**Pull request by tim-lai**
### Description

~Add a `process.env.NODE_ENV` check before logging a warning of a missing getComponent~
New: getComponent has a config object as a third, optional parameter
using the config object requires the presence of the second parameter, container
e.g. getComponent("JsonSchema_string_whatever", false, { failSilently: true })

JsonSchemaForm specifically now optionally suppresses the warning.

### Motivation and Context

extending #2931, where this is generally a safe warning and nothing needs to be done by the user. so we should silence the warning in production mode. 

### How Has This Been Tested?

tested manually using different NODE_ENV + CRA
dev mode, easiest to identify with query parameter, e.g. http://localhost:3200/?docExpansion=full

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [x] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [x] My changes can not or do not need to be tested.
- [ ] My changes can and should be tested by unit and/or integration tests.
- [ ] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/core/plugins/view/root-injects.jsx [hunk 1]
```diff
@@ -109,7 +109,11 @@ export const getComponent = (getSystem, getStore, getComponents, componentName,
   let component = getComponents(componentName)
 
   if(!component) {
-    getSystem().log.warn("Could not find component", componentName)
+    if (!process.env.NODE_ENV !== "production") {
+      // In the json schema rendering code, we optimistically query our system for a number of components.
+      // If the component doesn't exist, these warnings get thrown (in non-production mode).
+      getSystem().log.warn("Could not find component", componentName)
+    }
```
Comment (shockey) [inline src/core/plugins/view/root-injects.jsx]: Before we merge this @tim-lai, I'd like to go through the exercise of making sure that a consumer of Swagger UI that is in development is still able to see `getComponent` warnings. (They should, but it's certainly possible that something in our Webpack configs is transforming this value to be computed based on _our_ environment at build time, not theirs at runtime).

Roughly, here's how you'd do it:
1. Set up a project that is importing Swagger UI as a module, with a plugin that calls `getComponent` for a non-existing plugin. (My [swagger-ui-cra](https://github.com/shockey/swagger-ui-cra) repo is a bit outdated, but is a good starting point!)
2. Back in your Swagger UI folder, run `npm run build` while on this branch to generate fresh distribution assets that include your changes.
3. Use `npm link` to pull your freshly-generate assets into your project that consumes Swagger UI.
4. Run your project, and confirm that `getComponent` warnings are still visible!

**Comment by tim-lai** [comment]
please build

**Comment by tim-lai** [comment]
@shockey I had to fix a typo. Also squashed commits. That said, regarding 

> They should, but it's certainly possible that something in our Webpack configs is transforming this value to be computed based on our environment at build time, not theirs at runtime


Using a CRA, the `getComponent` warning is squashed. In comparison, SwaggerUI Dev mode does display the warning as expected.

For reference, here are my imports for CRA:
```
import SwaggerUI from 'swagger-ui';
import SwaggerUIStandalone from 'swagger-ui/dist/swagger-ui-standalone-preset';
```

**Comment by tim-lai** [comment]
@shockey new feature implemented... optional config parameter. JsonSchema specifically now optionally suppresses the warning. It works successfully in dev mode as well as CRA.

updating the description as well.

File: src/core/plugins/view/root-injects.jsx [hunk 2]
```diff
@@ -101,15 +101,20 @@ const wrapRender = (component) => {
 }
 
 
-export const getComponent = (getSystem, getStore, getComponents, componentName, container) => {
+export const getComponent = (getSystem, getStore, getComponents, componentName, container, config) => {
```
Comment (shockey) [inline src/core/plugins/view/root-injects.jsx]: **suggestion:** since this is a new optional parameter and we aren't updating existing call sites, let's set a sane default!
```suggestion
export const getComponent = (getSystem, getStore, getComponents, componentName, container, config = {}) => {
```

File: src/core/plugins/view/root-injects.jsx [hunk 3]
```diff
@@ -101,15 +101,20 @@ const wrapRender = (component) => {
...
     throw new TypeError("Need a string, to fetch a component. Was given a " + typeof componentName)
 
+    // getComponent has a config object as a third, optional parameter
+    // using the config object requires the presence of the second parameter, container
+    // e.g. getComponent("JsonSchema_string_whatever", false, { failSilently: true })
   let component = getComponents(componentName)
 
   if(!component) {
-    getSystem().log.warn("Could not find component", componentName)
+    if (!(config && config.failSilently)) {
```
Comment (shockey) [inline src/core/plugins/view/root-injects.jsx]: **suggestion:** in concert with the above, we can rely on a sane default value instead of doing maybe-object checks inline.

```suggestion
    if (!config.failSilently) {
```

File: src/core/plugins/view/root-injects.jsx [hunk 4]
```diff
@@ -101,15 +101,20 @@ const wrapRender = (component) => {
...
+      getSystem().log.warn("Could not find component:", componentName, " | config:", config)
```
Comment (shockey) [inline src/core/plugins/view/root-injects.jsx]: **suggestion:** I think logging the `config` object to users that are asking for a non-existent component is more likely to be noise than signal — doubly so in cases where `config` is omitted and is represented in the log line as `undefined`.

```suggestion
      getSystem().log.warn("Could not find component:", componentName)
```

File: src/core/json-schema-components.jsx [hunk 5]
```diff
@@ -50,8 +50,9 @@ export class JsonSchemaForm extends Component {
       schema = schema.toJS()
 
     let { type, format="" } = schema
-
-    let Comp = (format ? getComponent(`JsonSchema_${type}_${format}`) : getComponent(`JsonSchema_${type}`)) || getComponent("JsonSchema_string")
+    // In the json schema rendering code, we optimistically query our system for a number of components.
+    // If the component doesn't exist, we optionally suppress these warnings.
+    let Comp = (format ? getComponent(`JsonSchema_${type}_${format}`, false, { failSilently: true }) : getComponent(`JsonSchema_${type}`, false, { failSilently: true })) || getComponent("JsonSchema_string", false, { failSilently: true })
```
Comment (shockey) [inline src/core/json-schema-components.jsx]: **suggestion:** there's a _ton_ going on in L55 now, let's break out our common logic to a reusable function.

(let's follow the [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))!)

```suggestion
    // If the component doesn't exist, we optionally suppress these warnings.
    let getComponentSilently = (name) => getComponent(name, false, { failSilently: true })
    let Comp = (format ? getComponentSilently(`JsonSchema_${type}_${format}`) : getComponentSilently(`JsonSchema_${type}`)) || getComponentSilently("JsonSchema_string")
```

**Comment by shockey** [comment]
@tim-lai as you mentioned, docs would be great for this! I think a new example for this in the [Plugin API doc's _Components_ section](https://github.com/swagger-api/swagger-ui/blob/master/docs/customization/plugin-api.md#components) would be a good fit, perhaps something like:

```js
// use `config.failSilently` if you don't want a warning when
// a component doesn't exist in the system
const thisVariableWillBeNull = getComponent("not_real", false, { failSilently: true })
```
